File: problems/2901-longest-unequal-adjacent-groups-subsequence-ii/solution.cpp
// Link: https://leetcode.cn/problems/longest-unequal-adjacent-groups-subsequence-ii/
// Level: Medium
// Date: 2025-05-16

class Solution {
private:
    bool isAdjacent(const string& a, const string& b) {
        if (a.size() != b.size()) {
            return false;
        }
        int diff = 0;
        for (int i = 0; i < a.size(); i++) {
            if (a[i] != b[i]) {
                diff++;
                if (diff > 1) {
                    return false;
                }
            }
        }
        return diff == 1;
    }

public:
    vector<string> getWordsInLongestSubsequence(vector<string>& words, vector<int>& groups) {
        int n = words.size();
        vector<int> dp(n, 1);
        vector<int> prev(n, -1);
        int max_len = 1;
        int max_idx = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (groups[i] != groups[j] && isAdjacent(words[i], words[j])) {
                    if (dp[j] + 1 > dp[i]) {
                        dp[i] = dp[j] + 1;
                        prev[i] = j;
                    }
                }
            }
            if (dp[i] > max_len) {
                max_len = dp[i];
                max_idx = i;
            } else if (dp[i] == max_len) {
                max_idx = i;
            }
        }

        vector<int> path;
        int curr = max_idx;
        while (curr != -1) {
            path.push_back(curr);
            curr = prev[curr];
        }
        reverse(path.begin(), path.end());
        vector<string> res;
        for (int i : path) {
            res.push_back(words[i]);
        }
        return res;
    }
};

File: problems/2901-longest-unequal-adjacent-groups-subsequence-ii/solution.py
# Link: https://leetcode.cn/problems/longest-unequal-adjacent-groups-subsequence-ii/
# Level: Medium
# Date: 2025-05-16

class Solution:
    def getWordsInLongestSubsequence(self, words: List[str], groups: List[int]) -> List[str]:
        def is_adjacent(a: str, b: str) -> bool:
            if len(a) != len(b):
                return False
            diff = 0
            for x, y in zip(a, b):
                if x != y:
                    diff += 1
                    if diff > 1:
                        return False
            return diff == 1
        
        n = len(words)
        dp = [1] * n
        prev = [-1] * n
        max_len = 1
        max_idx = 0

        for i in range(n):
            for j in range(i):
                if groups[i] != groups[j] and is_adjacent(words[i], words[j]):
                    if dp[j] + 1 > dp[i]:
                        dp[i] = dp[j] + 1
                        prev[i] = j
            if dp[i] > max_len:
                max_len = dp[i]
                max_idx = i
            elif dp[i] == max_len:
                max_idx = i

        path = []
        curr = max_idx
        while curr != -1:
            path.append(words[curr])
            curr = prev[curr]
        path.reverse()
        res = []
        for i in range(len(path)):
            res.append(path[i])
        return res
